pkg/importer/vddk-datasource_amd64.go

```diff
@@ -855,35 +855,47 @@ func createVddkDataSource(endpoint string, accessKey string, secKey string, thum
 	// then get the list of changed blocks from VMware for a delta copy.
 	var changed *types.DiskChangeInfo
 	if currentSnapshot != nil && previousCheckpoint != "" {
+		// The QueryChangedDiskAreas returns maximally 2000 changed blocks. If the disk has more than 2000 blocks
+		// we need to query the next chunk of the blocks starting from previous
+		var lastChangedSize = 0
+		changed = &types.DiskChangeInfo{}
 		// Check if this is a snapshot or a change ID, and query disk areas as appropriate.
 		// Change IDs look like: 52 de c0 d9 b9 43 9d 10-61 d5 4c 1b e9 7b 65 63/81
 		changeIDPattern := `([0-9a-fA-F]{2}\s?)*-([0-9a-fA-F]{2}\s?)*\/([0-9a-fA-F]*)`
 		if matched, _ := regexp.MatchString(changeIDPattern, previousCheckpoint); matched {
-			request := types.QueryChangedDiskAreas{
-				ChangeId:    previousCheckpoint,
-				DeviceKey:   backingFileObject.Key,
-				Snapshot:    currentSnapshot,
-				StartOffset: 0,
-				This:        vmware.vm.Reference(),
-			}
-			response, err := QueryChangedDiskAreas(vmware.context, vmware.vm.Client(), &request)
-			if err != nil {
-				return nil, err
+			for changed.Length == 0 || lastChangedSize != 0 {
+				request := types.QueryChangedDiskAreas{
+					ChangeId:    previousCheckpoint,
+					DeviceKey:   backingFileObject.Key,
+					Snapshot:    currentSnapshot,
+					StartOffset: changed.Length,
+					This:        vmware.vm.Reference(),
+				}
+				response, err := QueryChangedDiskAreas(vmware.context, vmware.vm.Client(), &request)
+				if err != nil {
+					return nil, err
+				}
+				lastChangedSize = len(response.Returnval.ChangedArea)
+				changed.ChangedArea = append(changed.ChangedArea, response.Returnval.ChangedArea...)
```

Comment:
Are we not concerned about fetching all this data upfront and storing it in memory? What is the granularity of the change intervals?

Reply:
I think it's a concern, but not enough of one to rework this whole thing for this bug. I don't know exactly what the granularity is, but I doubt it is finer than the disk sector size. The structures in this list should be two int64s for starting offset and entry count, with a sub-list of change structures with two more int64s for starting offset and length. I'm sure there's a few more ints worth of overhead, but it's not huge amounts per entry.

We saw a single QueryChangedDiskAreas call hit 2000 entries in one request, so I would spitball something like 1MB per call? And it should only take a few calls to get all the changes. I'm sure we can find a pathological worst case to blow the number up, but probably most of the time more changes than this get rolled up into one bigger block.

Reply:
Would it be possible to get more data on this? Do a migration that was previously failing and log how many records and maybe average block size?

Reply:
Sure, we can probably even get the actual memory usage.

Reply:
I don't think it would be wise to let this grow unbounded, importer pods don't have a lot of memory